contracts/coconsensus/Coconsensus.sol

```diff
@@ -46,23 +47,310 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
+    mapping (bytes32 /* metachainId */ => ObserverI) public observers;
 
     /** Mapping of metachain id to the domain separators. */
-    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) domainSeparators;
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) public domainSeparators;
+
+
+    /* Modifiers */
+
+    modifier onlyRunningProtocore(bytes32 _metachainId)
+    {
+        require(
+            msg.sender == address(protocores[_metachainId]),
+            "Protocore is not available for the given metachain id."
+        );
+
+        _;
+    }
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Setup function does the initialization of all the mosaic
+     *         contracts on the auxiliary chain.
+     *
+     * @dev This function can be called only once.
+     */
+    function setup() public {
+
+        require(
+            originMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+        auxiliaryMetachainId = genesisAuxiliaryMetachainId;
+
+        /*
+         * Setup self protocore contract first. All other protocores will
+         * require the dynasty from the self protocore to finalise the genesis
+         * checkpoint.
+         */
+        setupProtocores(auxiliaryMetachainId);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+            /*
+             * The setup of self protocore is already done as a first step so
+             * skip the protocore setup if the current metachain id is equal to
+             * the auxiliary metachain id.
+             */
+            if (currentMetachainId != auxiliaryMetachainId) {
+                // Setup protocore contract for the given metachain id.
+                setupProtocores(currentMetachainId);
+            }
+
+            // Setup observer contract for the given metachain id.
+            setupObservers(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice finaliseCheckpoint() function finalises a checkpoint at
+     *         a metachain.
+     *
+     * @param _metachainId A metachain id to finalise a checkpoint.
+     * @param _blockNumber A block number of a checkpoint.
+     * @param _blockHash A block hash of a checkpoint.
+     *
+     * \pre `_metachainId` is not 0.
+     * \pre `_blockHash` is not 0.
+     * \pre `msg.sender` should be the protocore contract.
+     * \pre `_blockNumber` must be multiple of epoch length.
+     * \pre `_blockNumber` must be greater than the last finalized block number.
+     *
+     * \post Sets the `blockchains` mapping.
```

Comment:
I want us to have `\post` and `\pre` conditions a little bit more verbose: in a way that a reader does not need to need a code and can understand them in the context of public/external interfaces of function/contract.

Reply:
Updated.